File: android/app/src/test/java/bd/sicip/qavisit/pdf/BillHtmlTest.kt
// HTML builder is pure string assembly -- no Android/WebView needed, so structure bits are
// asserted directly against the generated markup here. the actual print-to-PDF path
// (pdf/BillPrinter.kt) needs a real WebView and is left to the manual/instrumented smoke
// check, not this suite (see BillLayoutTest.kt's old note, same rule applies).
package bd.sicip.qavisit.pdf

import bd.sicip.qavisit.domain.BillTotals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class BillHtmlTest {

    private fun leg(depDate: String, night: Int = 0, food: Double = 0.0, fare: Double = 100.0) = BillLeg(
        depDate = depDate, depTime = "10:00:00", depPlace = "A",
        arrDate = depDate, arrTime = "11:00:00", arrPlace = "B",
        mode = "Bus", travelClass = "AC", fare = fare, remarks = null,
        nightStay = night, foodDay = food,
    )

    @Test fun `same-day legs share one rowspanned date cell`() {
        val trip = BillTrip(
            purposeLine = "Purpose: X",
            legs = listOf(leg("2026-06-08", night = 1, food = 1.0), leg("2026-06-08"), leg("2026-06-09", night = 0, food = 0.5)),
            nights = 1, foodDays = 1.5,
        )
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(0.0, 0.0, 0.0, 0.0))

        // the two 2026-06-08 legs collapse into one rowspan="2" date cell (dep + arr date);
        // night/food merge over the WHOLE 3-leg trip (rowspan="3") and show the trip-level
        // values, not per-leg counts. (header's own rowspan="2" cells excluded via <tbody>.)
        val tbody = html.substringAfter("<tbody>")
        assertEquals(2, Regex("rowspan=\"2\"").findAll(tbody).count())
        assertTrue(tbody.contains("rowspan=\"3\">1</td>"))
        assertTrue(tbody.contains("rowspan=\"3\">1.50</td>"))
        // one leg row per leg, day-grouping only merges cells (rowspan), never drops a <tr>.
        assertEquals(3, Regex("<td class=\"place\">A</td>").findAll(html).count())
    }

    @Test fun `purpose line renders verbatim as the trip's band text`() {
        val trip = BillTrip(purposeLine = "Purpose: Capacity Assessment - AEOSIB (Ref: —, 6 May 2026)", legs = listOf(leg("2026-06-08")), nights = 0, foodDays = 0.0)
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(0.0, 0.0, 0.0, 0.0))

        assertTrue(html.contains("Purpose: Capacity Assessment - AEOSIB (Ref: —, 6 May 2026)"))
    }

    @Test fun `purposeDate prefers ref_date over start_date`() {
        assertEquals("6 May 2026", purposeDate(startDate = "2026-05-10", refDate = "2026-05-06"))
    }

    @Test fun `purposeDate falls back to start_date when ref_date is absent`() {
        assertEquals("10 May 2026", purposeDate(startDate = "2026-05-10", refDate = null))
    }

    @Test fun `fares render with thousands separators`() {
        val trip = BillTrip(purposeLine = "Purpose: X", legs = listOf(leg("2026-06-08", fare = 12345.0)), nights = 0, foodDays = 0.0)
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(ta = 12345.0, accommodation = 0.0, food = 0.0, net = 12345.0))

        assertTrue(html.contains("12,345"))
    }

    @Test fun `net claim row spells the amount in words`() {
        val html = buildBillHtml("Officer", "2026-06-10", emptyList(), BillTotals(ta = 0.0, accommodation = 0.0, food = 0.0, net = 1680.0))

        assertTrue(html.contains("Net claim bill TK. (In word) One Thousand Six Hundred Eighty Only"))
    }

    @Test fun `zero night and food values render as a dash`() {
        val trip = BillTrip(purposeLine = "Purpose: X", legs = listOf(leg("2026-06-08", night = 0, food = 0.0)), nights = 0, foodDays = 0.0)
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(0.0, 0.0, 0.0, 0.0))

        assertTrue(html.contains(">-</td>"))
    }

    @Test fun `zero fare leg renders as a dash`() {
        val trip = BillTrip(purposeLine = "Purpose: X", legs = listOf(leg("2026-06-08", fare = 0.0)), nights = 0, foodDays = 0.0)
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(0.0, 0.0, 0.0, 0.0))

        assertTrue(html.contains("<td class=\"money\">-</td>"))
    }

    @Test fun `N slash A mode leg prints a dash in the mode cell`() {
        // real N/A legs carry a null class too (see ui.bill.BillScreen.toBillTrip) -- both the
        // mode and class cells should print bare "-" (no rowspan/class attrs on either).
        val na = BillLeg(
            depDate = "2026-06-08", depTime = "10:00:00", depPlace = "A",
            arrDate = "2026-06-08", arrTime = "11:00:00", arrPlace = "B",
            mode = "N/A", travelClass = null, fare = 100.0, remarks = "note",
            nightStay = 1, foodDay = 1.0,
        )
        val trip = BillTrip(purposeLine = "Purpose: X", legs = listOf(na), nights = 1, foodDays = 1.0)
        val html = buildBillHtml("Officer", "2026-06-10", listOf(trip), BillTotals(0.0, 0.0, 0.0, 0.0))

        assertEquals(2, Regex("<td>-</td>").findAll(html.substringAfter("<tbody>")).count())
    }
}

File: web/src/lib/billhtml.test.js
// buildBillHtml -- ported from android BillHtmlTest.kt (date/time formatting, day-grouping
// rowspans, dash-for-zero). string-shape checks only -- the actual print/PDF layer is browser
// window.print(), not JVM-testable here.
import { describe, it, expect } from 'vitest'
import { buildBillHtml, buildLocalBillHtml, localBillTrips } from './billhtml.js'
import { TICKET_REMARK } from './seeds.js'

const totals = { ta: 623, accommodation: 12000, food: 9750, net: 22373 }

function trip(overrides = {}) {
  return {
    purposeLine: 'Monitoring Visit - BGMEA (Ref: 123, 01 Jun 2026)',
    nights: 6, foodDays: 6.5,
    legs: [
      { depDate: '2026-06-08', depTime: '09:00', depPlace: 'Dhaka', arrDate: '2026-06-08', arrTime: '13:00', arrPlace: 'Bhola', mode: 'Bus', travelClass: 'AC', fare: 441, remarks: '', nightStay: 1, foodDay: 1 },
      { depDate: '2026-06-08', depTime: '14:00', depPlace: 'Bhola town', arrDate: '2026-06-08', arrTime: '14:30', arrPlace: 'Institute', mode: 'CNG', travelClass: 'Rented', fare: 50, remarks: '', nightStay: 0, foodDay: 0 },
      { depDate: '2026-06-14', depTime: '17:00', depPlace: 'Bhola', arrDate: '2026-06-14', arrTime: '21:00', arrPlace: 'Dhaka', mode: 'Bus', travelClass: 'AC', fare: 482, remarks: '', nightStay: 0, foodDay: 0.5 },
    ],
    ...overrides,
  }
}

it('header shows dd MMM yyyy date and officer name/designation', () => {
  const html = buildBillHtml('Mahfuzul Islam', '2026-07-01', [trip()], totals)
  expect(html).toContain('<b>Date:</b> 01 Jul 2026')
  expect(html).toContain('Mahfuzul Islam, Program Officer (QA)')
})

it('purpose band prefixes Purpose: and escapes html', () => {
  const html = buildBillHtml('X', '2026-07-01', [trip({ purposeLine: 'A & B <script>' })], totals)
  expect(html).toContain('Purpose: A &amp; B &lt;script&gt;')
})

it('itinerary dates render dd-MMM-yy and times render h:mm a', () => {
  const html = buildBillHtml('X', '2026-07-01', [trip()], totals)
  expect(html).toContain('08-Jun-26')
  expect(html).toContain('9:00 AM')
  expect(html).toContain('2:00 PM')
})

it('day-group shares rowspan=2 date cells; night/food merge over the whole trip', () => {
  const html = buildBillHtml('X', '2026-07-01', [trip()], totals)
  const tbody = html.slice(html.indexOf('<tbody>'))
  // one 2-leg day-group -> exactly 2 rowspan=2 cells (dep date, arr date); night/food span
  // the whole 3-leg trip and show trip-level values, not per-leg counts.
  const spanCells = (tbody.match(/rowspan="2"/g) || []).length
  expect(spanCells).toBe(2)
  expect(tbody).toContain('rowspan="3">6</td>')
  expect(tbody).toContain('rowspan="3">6.50</td>')
})

it('zero night/food cells render as dash', () => {
  const html = buildBillHtml('X', '2026-07-01', [trip({ nights: 0, foodDays: 0, legs: trip().legs.map((l) => ({ ...l, nightStay: 0, foodDay: 0 })) })], totals)
  expect(html).toMatch(/rowspan="3">-<\/td>/)
})

it('totals row sums nights/food across trips and shows net in words', () => {
  const html = buildBillHtml('X', '2026-07-01', [trip(), trip()], { ta: 623, accommodation: 24000, food: 19500, net: 44123 })
  expect(html).toContain('<b>12</b>') // nights summed (6+6)
  expect(html).toContain('Forty Four Thousand One Hundred Twenty Three')
})

describe('logos and print', () => {
  it('references /logos/*.jpg (public/, not android_asset)', () => {
    const html = buildBillHtml('X', '2026-07-01', [trip()], totals)
    expect(html).toContain('/logos/bd-govt-seal.jpg')
    expect(html).toContain('/logos/sicip-logo.jpg')
  })
})

it('zero fare renders as dash in leg and totals cells', () => {
  const zeroTrip = trip({ legs: trip().legs.map((l) => ({ ...l, fare: 0 })) })
  const html = buildBillHtml('X', '2026-07-01', [zeroTrip], { ta: 0, accommodation: 0, food: 0, net: 0 })
  expect(html).toMatch(/class="money">-<\/td>/)
  expect(html).toMatch(/class="money"><b>-<\/b><\/td>/)
})

function localLeg(overrides = {}) {
  return {
    depDate: '2026-06-08', depTime: '09:00', depPlace: 'Dhaka', arrDate: '2026-06-08', arrTime: '10:00', arrPlace: 'Bhola',
    mode: 'CNG', fare: 50, remarks: '',
    ...overrides,
  }
}

describe('localBillTrips', () => {
  it('drops zero-fare and ticket-remark legs; keeps the rest', () => {
    const trips = [{
      purposeLine: 'P1',
      legs: [
        localLeg({ fare: 0 }),
        localLeg({ remarks: TICKET_REMARK }),
        localLeg({ fare: 100 }),
      ],
    }]
    const out = localBillTrips(trips)
    expect(out.length).toBe(1)
    expect(out[0].legs.length).toBe(1)
    expect(out[0].legs[0].fare).toBe(100)
  })

  it('N/A-mode leg with zero fare is still dropped by the fare rule', () => {
    const trips = [{ purposeLine: 'P1', legs: [localLeg({ mode: 'N/A', fare: 0 })] }]
    expect(localBillTrips(trips)).toEqual([])
  })

  it('N/A-mode leg with fare and no ticket remark is kept and prints dash mode', () => {
    const trips = [{ purposeLine: 'P1', legs: [localLeg({ mode: 'N/A', fare: 150 })] }]
    const out = localBillTrips(trips)
    expect(out[0].legs.length).toBe(1)

    const html = buildLocalBillHtml('X', '2026-07-01', trips)
    expect(html).toContain('<td>-</td>')
  })

  it('drops a trip whose every leg is filtered out', () => {
    const trips = [{ purposeLine: 'AllGone', legs: [localLeg({ fare: 0 }), localLeg({ fare: 0, mode: 'N/A' })] }]
    expect(localBillTrips(trips)).toEqual([])
  })
})

describe('buildLocalBillHtml', () => {
  const localTrips = [{
    purposeLine: 'Local errands',
    legs: [
      localLeg({ fare: 100 }),
      localLeg({ fare: 0 }), // filtered out
      localLeg({ fare: 50, depDate: '2026-06-09', arrDate: '2026-06-09' }),
    ],
  }]

  it('renders the local band and omits Recommended By', () => {
    const html = buildLocalBillHtml('X', '2026-07-01', localTrips)
    expect(html).toContain('Detailed Local Travel Itinerary:')
    expect(html).not.toContain('Recommended By')
  })

  it('total row sums only surviving legs\' fare', () => {
    const html = buildLocalBillHtml('X', '2026-07-01', localTrips)
    expect(html).toContain('<b>150</b>') // 100 + 50, zero-fare leg dropped
  })
})
